### PR TITLE
Update documentWrapper to have the error property

### PR DESCRIPTION
**Pull request by smithab**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests if this modifies the UWP implementation?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

The DocumentWrapper needs to have a DataException captured. When we deserialize a document and it fails we need to return an error object. 

## Related PRs or issues

List related PRs and other issues.

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

File: SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/ConversionExtensions.cs [hunk 1]
```diff
@@ -15,19 +15,30 @@ public static class ConversionExtensions
...
                 ETag = documentWrapper.ETag,
                 Id = documentWrapper.DocumentId,
-                LastUpdatedDate = documentWrapper.LastUpdatedDate != null ? (DateTime)documentWrapper?.LastUpdatedDate : UnixEpoch,
                 IsFromDeviceCache = documentWrapper.FromDeviceCache
             };
-        }
+
+            try
+            {
+                var jsonValue = documentWrapper.JsonValue;
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/ConversionExtensions.cs]: why is jsonValue inside the try/catch block? we should expose it even if deserializing fails.

File: SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/ConversionExtensions.cs [hunk 2]
```diff
@@ -15,19 +15,30 @@ public static class ConversionExtensions
...
+                var deserializedValue = jsonValue != null ? JsonConvert.DeserializeObject<T>(jsonValue) : default(T);
+                parsedDocument.DeserializedValue = deserializedValue;
+                parsedDocument.JsonValue = jsonValue;
+                parsedDocument.LastUpdatedDate = documentWrapper.LastUpdatedDate != null ? (DateTime)documentWrapper?.LastUpdatedDate : UnixEpoch;
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/ConversionExtensions.cs]: I don't understand why the LastUpdatedDate has 2 variants here from try or catch.

The only exception that can happen here is newtonsoft json deserialization.

File: SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/ConversionExtensions.cs [hunk 3]
```diff
@@ -15,19 +15,30 @@ public static class ConversionExtensions
...
-                DeserializedValue = deserializedValue,
-                JsonValue = jsonValue,
                 Partition = documentWrapper.Partition,
                 ETag = documentWrapper.ETag,
                 Id = documentWrapper.DocumentId,
-                LastUpdatedDate = documentWrapper.LastUpdatedDate != null ? (DateTime)documentWrapper?.LastUpdatedDate : UnixEpoch,
                 IsFromDeviceCache = documentWrapper.FromDeviceCache
             };
-        }
+
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.iOS/ConversionExtensions.cs]: don't put a new line if there is no comment

File: SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/DocumentWrapper.cs [hunk 4]
```diff
@@ -21,6 +21,12 @@ public class DocumentWrapper<T> : DocumentMetadata
         /// </summary>
         public string JsonValue { get; internal set; }
 
+        /// <summary>
+        /// The document error, or null if the operation was successful.
+        /// </summary>
+        /// <value>The error.</value>
+        public DataException Error { get; internal set; }
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/DocumentWrapper.cs]: We need the Android change too

**Comment by guperrot** [comment]
sounds like #1013 replaces this PR

File: SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/ConversionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            // Try to deserialize.
+            try
+            {
+                // Set deserialized value.
+                if (parsedDocument.JsonValue != null)
+                {
+                    parsedDocument.DeserializedValue = deserealizer(parsedDocument.JsonValue);
+                }
+            }
+            catch (Exception ex)
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/ConversionExtensions.cs]: Can it throw anything else than JsonException actually? We should not catch knives.

Comment (tjyung) [reply SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/ConversionExtensions.cs]: We aren't able to reference Newtonsoft.json in the shared project and therefore were not able to use a JsonException.

Comment (guperrot) [reply SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/ConversionExtensions.cs]: ok

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/ModulePages/OthersContentPage.xaml.cs [hunk 6]
```diff
@@ -116,7 +116,7 @@ async void RunMBaaSAsync(object sender, EventArgs e)
             }
             try
             {
-                var list = await Data.ListAsync<CustomDocument>(DefaultPartitions.UserDocuments);
+                var list = await Data.ListAsync<CustomDocument>(DefaultPartitions.UserDocuments, new ReadOptions());
```
Comment (guperrot) [inline Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/ModulePages/OthersContentPage.xaml.cs]: Does not seem related to the changes in this PR, we should probably not commit that.

File: SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/ConversionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
+﻿using System;
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/ConversionExtensions.cs]: Copyright missing.

File: SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/Microsoft.AppCenter.Data.Shared.projitems [hunk 8]
```diff
@@ -21,5 +21,6 @@
     <Compile Include="$(MSBuildThisFileDirectory)DocumentMetadata.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)DefaultPartitions.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)RemoteOperationCompletedEventArgs.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)ConversionExtensions.cs" />
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/Microsoft.AppCenter.Data.Shared.projitems]: Please rename the file as the content is a different class than the file name, it's confusing.

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/ModulePages/OthersContentPage.xaml.cs (hunk 6 above)
Comment (jwallra) [reply Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/ModulePages/OthersContentPage.xaml.cs]: I think it's an OK fix, since the API supports it.  No harm done in this case since there is no functionality change

File: SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/ConversionExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Microsoft.AppCenter.Data
+{
+    public static class SharedConversionExtensions
+    {
+        public static DocumentWrapper<T> ToDocumentWrapper<T>(
```
Comment (jwallra) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/ConversionExtensions.cs]: nit: We don't usually format function signatures like this, just put them in line and then wrap when the line gets too long

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/ModulePages/OthersContentPage.xaml.cs (hunk 6 above)
Comment (guperrot) [reply Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/ModulePages/OthersContentPage.xaml.cs]: But there is no fix here.

File: SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/SharedConversionExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+            {
+                // Set deserialized value.
+                if (parsedDocument.JsonValue != null)
+                {
+                    parsedDocument.DeserializedValue = deserealizer(parsedDocument.JsonValue);
+                }
+            }
+            catch (Exception ex)
+            {
+                parsedDocument.Error = new DataException(ex.Message);
```
Comment (guperrot) [inline SDK/AppCenterData/Microsoft.AppCenter.Data.Shared/SharedConversionExtensions.cs]: DocumentMetadata should be set on the Error as well.
